accounts/accounts/controllers/registration.py

```diff
@@ -0,0 +1,303 @@
+                                choices=GROUPS, default='')
+    default_category = OptGroupSelectField('Your default category',
+                                           choices=CATEGORIES, default='')
+
+    url = StringField('Your homepage URL', validators=[optional(),
+                      Length(max=255), URL()])
+    remember_me = BooleanField('Have your browser remember who you are?',
+                               default=True)
+
+    @classmethod
+    def from_domain(cls, user: domain.User) -> 'ProfileForm':
+        """Instantiate this form with data from a domain object."""
+        return cls(MultiDict({  # type: ignore
+            'username': user.username,
+            'email': user.email,
+            'forename': user.name.forename,
+            'surname': user.name.surname,
+            'suffix': user.name.suffix,
+            'organization': user.profile.organization,
+            'country': user.profile.country.upper(),
+            'status': user.profile.rank,
+            'groups': user.profile.submission_groups,
+            'default_category': user.profile.default_category.compound,
+            'url': user.profile.homepage_url,
+            'remember_me': user.profile.remember_me
+        }))
+
+    def to_domain(self) -> domain.User:
+        """Generate a :class:`.User` from this form's data."""
+        return domain.User(
+            user_id=self.user_id.data if self.user_id.data else None,
+            username=self.username.data,
+            email=self.email.data,
+            name=domain.UserFullName(
+                forename=self.forename.data,
+                surname=self.surname.data,
+                suffix=self.suffix.data
+            ),
+            profile=domain.UserProfile(
+                organization=self.organization.data,
+                country=self.country.data,
+                rank=int(self.status.data),     # WTF can't handle int values.
+                submission_groups=self.groups.data,
+                default_category=domain.Category(
+                    *self.default_category.data.split('.')
+                ),
+                homepage_url=self.url.data,
+                remember_me=self.remember_me.data
+            )
+        )
+
+    def validate_username(self, field: StringField) -> None:
+        """Ensure that the username is unique."""
+        if users.username_exists(field.data):
+            raise ValidationError('An account with that username already'
+                                  ' exists')
+
+    def validate_email(self, field: StringField) -> None:
+        """Ensure that the email address is unique."""
+        if users.email_exists(field.data):
+            raise ValidationError('An account with that email address'
+                                  ' already exists')
+
+
+class RegistrationForm(ProfileForm):
+    """User registration form."""
+
+    password = PasswordField('Password',
```

Comment:
I think we'll want to support higher password entropy here. The legacy system has an 8-character minimum and enforces a number of validation rules.